bpf/lib/nodeport.h

```diff
@@ -1403,10 +1465,35 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
 
 	return 0;
 }
+#elif DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
+static __always_inline int dsr_encap_geneve_opt4(struct __ctx_buff *ctx,
+						 struct iphdr *ip4, __be32 svc_addr,
+						 __be16 svc_port, int *ifindex)
+{
+	struct remote_endpoint_info *info = NULL;
+	struct geneve_opt4 gopt;
+
+	info = ipcache_lookup4(&IPCACHE_MAP, ip4->daddr, V4_CACHE_KEY_LEN, 0);
+	if (info && info->tunnel_endpoint != 0) {
+		set_geneve_opt4(svc_port, svc_addr, &gopt);
+
+		return  __encap_with_nodeid_opt4(ctx, info->tunnel_endpoint,
+						 WORLD_ID,
```

Comment:
Ah, this is interesting. I can see how this would obviously work when in TUNNEL_MODE. But we also expect to run with direct-routing, correct? And only establish the tunnel-infrastructure for DSR, so that we can encapsulate. Would the `tunnel_endpoint` be populated in that case?

(we've only used this tunnel-with-direct-routing-mode for EgressGW so far, and there we get the tunnel_endpoint IP from the EgressGW policy entry ...)

Reply:
Yes, exactly. `tunnel_endpoint` will be populated with direct-routing mode as well, and it works